src/main/java/org/truffleruby/language/methods/DeclarationContext.java

```diff
@@ -89,7 +90,7 @@ public DeclarationContext(
             Map<DynamicObject, DynamicObject[]> refinements) {
         this.visibility = visibility;
         this.defaultDefinee = defaultDefinee;
-        this.refinements = refinements;
+        this.refinements = refinements == null || refinements.isEmpty() ? NO_REFINEMENTS : refinements;
```

Comment:
This fails the methods in PE paths check:
```
=== Found 1 compilation blacklist violations ===

Blacklisted method
  java.util.HashMap.isEmpty()
called from
  org.truffleruby.language.methods.DeclarationContext.<init>(DeclarationContext.java:93)
  org.truffleruby.core.basicobject.BasicObjectNodes$InstanceExecNode.instanceExec(BasicObjectNodes.java:404)
  org.truffleruby.core.basicobject.BasicObjectNodesFactory$InstanceExecNodeFactory$InstanceExecNodeGen.execute(BasicObjectNodesFactory.java:1349)
  org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:60)
  org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:484)
  org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:461)
```

I'd suggest to just assign here, and `assert refinements == NO_REFINEMENTS || !refinements.isEmpty();`

Reply:
The assert check works for me :+1: 

Btw, what is `PE paths check` ? Can I run this locally?

Reply:
Hm, I hurried:

https://github.com/oracle/truffleruby/blob/443e77af313d6b550e17398c0f31b6277ee69672/src/main/java/org/truffleruby/language/methods/UsingNode.java#L51

If we have empty module:
```ruby
it "accepts module as argument" do
    refinement = Module.new do
      refine Integer do
        def foo; "foo"; end
      end
    end

    -> {
      Module.new do
        using refinement
      end
    }.should_not raise_error
  end
```

we will get

```
Module#using accepts module without refinements FAILED
Expected to not get Exception
but got: RuntimeError (<no message> (AssertionError
```

because refinements is empty `HashMap` but not `NO_REFINEMENTS`

Reply:
>  Can I run this locally?

Yes, just `jt build --env native` should reproduce it. It's the `-H:+TruffleCheckBlackListedMethods` flag that shows this, which we have on by default.

I'm a bit confused about that empty case, I'll reproduce it locally.

Reply:
> I'm a bit confused about that empty case, I'll reproduce it locally.

AFAIU `final Map<DynamicObject, DynamicObject[]> newRefinements = new HashMap<>(declarationContext.getRefinements());` doesn't changed when module is empty and `newRefinements` passed as is.

See: https://github.com/ruby/spec/blob/ffb824547923b0dfa1dba135453da8b4d0484019/core/module/using_spec.rb#L20-L32

Reply:
I see, UsingNode might have an empty `newRefinements`.
I think the best here is to do this:
```java
        if (!newRefinements.isEmpty()) {
            DeclarationContext.setRefinements(callerFrame, declarationContext, newRefinements);
        }
```

Reply:
Hm, looks like it works but I don't get it how :D

From my point of view, `newRefinements.isEmpty())` is the same 
```
Blacklisted method
  java.util.HashMap.isEmpty()
```

p.s. I still think it's better to handle edge cases inside the constructor, rather than relying on external handlers (like `if` above). But you are the boss here :)

Reply:
The issue here is the constructor is called on the fast-path, and Map#isEmpty is a virtual call. So the constructor should be fast and ideally branchless.
OTOH, the only caller that can give an empty HashMap (which is very suboptimal, it will cause hash code computations for nothing, etc) is behind a TruffleBoundary, so performance there is far less critical.

Reply:
Thanks so much for the detailed explanation! And sorry for the dummy questions - I like to know the reason for the changes, but not just do it :)